mne/simulation/metrics/metrics.py

```diff
@@ -0,0 +1,562 @@
+                         (len(stc1.vertices), len(stc2.vertices)))
+    idx_start1 = 0
+    idx_start2 = 0
+    stc1 = stc1.copy()
+    stc2 = stc2.copy()
+    all_data1 = []
+    all_data2 = []
+    for i, (vert1, vert2) in enumerate(zip(stc1.vertices, stc2.vertices)):
+        vert = np.union1d(vert1, vert2)
+        data1 = np.zeros([len(vert), stc1.data.shape[1]])
+        data2 = np.zeros([len(vert), stc2.data.shape[1]])
+        data1[np.searchsorted(vert, vert1)] = \
+            stc1.data[idx_start1:idx_start1 + len(vert1)]
+        data2[np.searchsorted(vert, vert2)] = \
+            stc2.data[idx_start2:idx_start2 + len(vert2)]
+        idx_start1 += len(vert1)
+        idx_start2 += len(vert2)
+        stc1.vertices[i] = vert
+        stc2.vertices[i] = vert
+        all_data1.append(data1)
+        all_data2.append(data2)
+
+    stc1._data = np.concatenate(all_data1, axis=0)
+    stc2._data = np.concatenate(all_data2, axis=0)
+    return stc1, stc2
+
+
+def _apply(func, stc_true, stc_est, per_sample):
+    """Apply metric to stcs.
+
+    Applies a metric to each pair of columns of stc_true and stc_est
+    if per_sample is True. Otherwise it applies it to stc_true and stc_est
+    directly.
+    """
+    if per_sample:
+        metric = np.empty(stc_true.data.shape[1])  # one value per time point
+        for i in range(stc_true.data.shape[1]):
+            metric[i] = func(stc_true.data[:, i:i + 1],
+                             stc_est.data[:, i:i + 1])
+    else:
+        metric = func(stc_true.data, stc_est.data)
+    return metric
+
+
+def _thresholding(stc_true, stc_est, threshold):
+    if isinstance(threshold, str):
+        t = _check_threshold(threshold)
+        if stc_true is not None:
+            stc_true._data[np.abs(stc_true._data) <=
+                           t * np.max(np.abs(stc_true._data))] = 0.
+        stc_est._data[np.abs(stc_est._data) <=
+                      t * np.max(np.abs(stc_est._data))] = 0.
+    else:
+        if stc_true is not None:
+            stc_true._data[np.abs(stc_true._data) <= threshold] = 0.
+        stc_est._data[np.abs(stc_est._data) <= threshold] = 0.
+    return stc_true, stc_est
+
+
+def _cosine(x, y):
+    p = np.reshape(x, (-1, 1))
+    q = np.reshape(y, (-1, 1))
+    if norm(p) * norm(q):
+        return (np.dot(p.T, q) / (norm(p) * norm(q)))[0][0]
+    elif norm(p) == norm(q):
+        return 1
+    else:
+        return 0
+
+
+@fill_doc
+def cosine_score(stc_true, stc_est, per_sample=True):
+    """Compute cosine similarity between 2 source estimates.
+
+    Parameters
+    ----------
+    %(metric_stc_true)s
+    %(metric_stc_est)s
+    %(metric_per_sample)s
+
+    Returns
+    -------
+    %(stc_metric)s
+    """
+    stc_true, stc_est = _uniform_stc(stc_true, stc_est)
+    func = partial(_cosine)
+    metric = _apply(func, stc_true, stc_est, per_sample=per_sample)
+    return metric
+
+
+def _check_threshold(threshold):
```

Comment:
this function does not do what it says it does! it does not check that threshold is a float.
maybe misses `_check_option('threshold', threshold, (float, str))` and we never make sure that threshold is in `[0,1]`

Reply:
agreed